### ISSUE TITLE
Cannot stub Engine classes

### ISSUE THREAD

**Issue by Razzlegames** [bitwes/Gut#133]
Hi,

Great framework, I'm a bit new and I think I've either ran into a current limitation or a misunderstanding my my part.  I currently cannot seem to mock out Engine objects, for example `RayCast2D` objects.

In this example, I'd like to stub out calls to `topRay.get_collider()`, to return a physics body I provide in the test.

When I do so I get the error `Invalid get index '_gut_metadata_`(on base: 'RayCast2D').`

```gdscript
extends "res://addons/gut/test.gd"

var enemyCollisionBehavior

func before_each():
	enemyCollisionBehavior = load(
		"res://Prefabs/Characters/Behaviors/EnemyCollisionBehavior.tscn").instance()

func test_objectOnTop():
	var body = Node.new()
	
	var topRay = RayCast2D.new()
	stub(topRay, 'get_collider').to_return(body)
```

Any ideas on what I could be doing wrong or can we only mock/stub out custom implementations?

**Comment by Razzlegames** [comment]
![gut_mocking_error](https://user-images.githubusercontent.com/8754559/68828034-61c33980-0661-11ea-80ec-3eef2b8f0629.png)
Error message

**Comment by Razzlegames** [comment]
Actually I read this:

>Because of how doubling works, you can only stub methods that are implemented in your script or any scripts that inherits from. You can stub most built-in methods if you use the experimental FULL doubling strategy. Check the bottom of the doubling page for more information.

But I can't seem to locate the docs on the experimental features.

By the way, due to duck typing I can get around the limitation but it is a bit cumbersome to define a new class for every mocking use case.


```gdscript

extends "res://addons/gut/test.gd"

var enemyCollisionBehavior

func before_each():
	enemyCollisionBehavior = load(
		"res://Prefabs/Characters/Behaviors/EnemyCollisionBehavior.tscn").instance()

func test_objectOnTop():
	var body = Node.new()
	
	var topRay = RayCast2DMock.new()
	topRay.body = body
		
	enemyCollisionBehavior.setTopRays([topRay])
	var result = enemyCollisionBehavior.objectOnTop(body)
	assert_true(result)
	
func test_objectNotOnTop():
	var body = Node.new()
	
	var topRay = RayCast2DMock.new()
	topRay.body = body
		
	enemyCollisionBehavior.setTopRays([topRay])
	var result = enemyCollisionBehavior.objectOnTop(Node.new())
	assert_false(result)
	
class RayCast2DMock:
	var body
	func get_collider():
		return body
```

**Comment by bitwes** [comment]
It looks like you are missing a call to `double`.  You can only stub things that have been doubled first.  Check the wiki page for Doubles.  The experimental doubling strategy is also at the bottom of that page.

I've not double raycasts methods so let me know if it works.

**Comment by Razzlegames** [comment]
Thanks for getting back to me.

Ah I see the experimental strategy is on the `double` document page.  Just FYI the Stubbing page reads:

>  experimental FULL doubling strategy. Check the bottom of the doubling page for more information.

But I think that's a mistake since it's not there?
____

On the `Doubles` page: https://github.com/bitwes/Gut/wiki/Doubles I noticed "You can spy and stub most of the Built-Ins in Godot", but I am having trouble finding a good example of how to do this. 

I've tried `double` this prior to posting, but thought maybe I was doing something wrong since all I get are errors

```gdscript
var mockRayCast2D = double(RayCast2D.new(), DOUBLE_STRATEGY.FULL)
```
Error is Invalid get index

```gdscript
	set_double_strategy(DOUBLE_STRATEGY.FULL)
	var mockRayCast2D = double(RayCast2D.new())
```
Same...

```
	set_double_strategy(DOUBLE_STRATEGY.FULL)
	var mockRayCast2D = double(RayCast2D)
``` 
No go :(

Any idea?

**Comment by bitwes** [comment]
I just double checked and that isn't going to work.  `RayCast2D` is a `GDScriptNativeClass` and GUT doesn't support that.  It's kinda like trying to double an `int` as far as GUT is concerned.  I did this:
```
extends "res://addons/gut/test.gd"

func test_ray_cast():
	var body = Node.new()
	var topRay = partial_double(RayCast2D).new()
	stub(topRay, 'get_collider').to_return(body)
```
and got:
```
Debugger Break, Reason: 'Invalid get index 'resource_path' (on base: 'GDScriptNativeClass').'
*Frame 0 - res://addons/gut/test.gd:124 in function '_init'
Enter "help" for assistance.
```
I could probably detect this and give a better message.  The test will probably still blow up b/c the thing wouldn't have been doubled but at least the output could lead you in the right direction.

You can double things like a `Node2D` or a `Button` but `RayCast` is a little more low level than those it appears.\

Also note that you should do `double(thing).new()` and not `double(thing.new())`.  `double` needs the "class" not an instance.  It looks at the "class" definition and creates a fake one with a bunch of stubbed out methods.

I'm going to leave this open and see if there isn't something that can be done.  I'll at least create a different bug to put out an error message if nothing else.

**Comment by bitwes** [comment]
Putting this here so I don't forget.  It looks like I might be able to check for a `GDScriptNativeClass` in the constructor of the `DoubleInfo` helper class in `test.gd`.  This could be the error message or it could possibly be a new double path.  

The problem I see is that GUT uses the `resource_path` and extension that it derives from that to figure out if the thing is a scene or a script.  It also uses this information to identify the object in the `stubber`.  It seems feasible to be able to double/stub `GDScriptNativeClass` but it might take some rework to get it done.

**Comment by Razzlegames** [comment]
>  It seems feasible to be able to double/stub GDScriptNativeClass but it might take some rework to get it done.

I haven't dug deep into the `GUT` internals, but I was thinking it would be nice to be able to generate duck typed mocks, for any type, from just reflectively looking up the methods available in an object... then generating the stub with all of these methods (which by default either `fail` or are `pass` noops by default (depending on preference). 

 When stubbing the method with an implementation we could generate the classes.  It's too bad though I don't know any way to add method definitions at run time in `gdscript`, unlike other reflective languages.  This may be down a rabbit hole a bit. 

 The other solution is actually generating code for the duck typed mock (in temp folders or something) but that's a mess as it should be removed after ever test run etc (probably).

**Comment by Razzlegames** [comment]
Oh I see you're already doing that... interesting.  I guess if they ever add reflective `obj.add_method("name", someLambda)` it will make our lives easier, I think ;)

```gdscript
func _write_file(obj_info, dest_path, override_path=null):
	var script_methods = _get_methods(obj_info)

	var metadata = _get_stubber_metadata_text(obj_info)
	if(override_path):
		metadata = _get_stubber_metadata_text(obj_info, override_path)

	var f = File.new()
	f.open(dest_path, f.WRITE)

	f.store_string(str(obj_info.get_extends_text(), "\n"))
	f.store_string(metadata)

	for i in range(script_methods.local_methods.size()):
		if(obj_info.make_partial_double):
			_stub_to_call_super(obj_info, script_methods.local_methods[i].name)
		f.store_string(_get_func_text(script_methods.local_methods[i]))

	for i in range(script_methods.built_ins.size()):
		_stub_to_call_super(obj_info, script_methods.built_ins[i].name)
		f.store_string(_get_func_text(script_methods.built_ins[i]))

	f.close()
```

**Comment by bitwes** [comment]
Yea, if we had a way to monkey patch in gdscript the classes could be generated without having to make the temp files.  I've thought about switching over to not write files.  I created a POC of it here:  https://github.com/bitwes/Gut/blob/master/scratch/make_class_from_string.gd.

The downside is that when adding new functionality to the generated code it is super nice to be able to look at a file when syntax errors occur.  Dynamically generating code can get quite ugly.  I could create debug files before executing, but it's been low on the list.  I think there might be some cleanup bugs too, so if anyone finds those and complains that might be the push to go to using strings instead of files.

I'm going to look into doubling built-ins.  as a workaround you could create a class that inherits from `RayCast2D` and then `double` that.  That should work.  You don't have to do anything in it, just `extend RayCast2D`.

**Comment by bitwes** [comment]
Sorry, you will have to create a class that extends `RayCast2D` and double using `DOUBLE_STRATEGY.FULL`.  Without the `DOUBLE_STRATEGY.FULL` you will only get methods you define in your subclass.

**Comment by bitwes** [comment]
I've got it working at a low level.  I've got some work to do for stubbing and I haven't tested spies yet, but it looks like this will make it into the next release.

**Comment by bitwes** [comment]
Implemented in 6.8.1

**Comment by Razzlegames** [comment]
Thanks man! I'll have to give it a go :)
